pkg/scanners/scannerv4/convert.go

```diff
@@ -50,13 +50,20 @@ func components(metadata *storage.ImageMetadata, report *v4.VulnerabilityReport)
 		}
 
 		component := &storage.EmbeddedImageScanComponent{
-			Name:          pkg.GetName(),
-			Version:       pkg.GetVersion(),
-			Vulns:         vulnerabilities(report.GetVulnerabilities(), vulnIDs),
-			FixedBy:       pkg.GetFixedInVersion(),
-			Source:        source,
-			Location:      location,
-			HasLayerIndex: layerIdx,
+			Name:     pkg.GetName(),
+			Version:  pkg.GetVersion(),
+			Vulns:    vulnerabilities(report.GetVulnerabilities(), vulnIDs),
+			FixedBy:  pkg.GetFixedInVersion(),
+			Source:   source,
+			Location: location,
+		}
+		// DO NOT BLINDLY SET THIS INSIDE THE STRUCT DECLARATION DIRECTLY ABOVE.
+		// IF layerIdx IS nil, IT DOES NOT MEAN HasLayerIndex WILL BE THE SAME nil.
+		// GO CAN SOMETIMES BE ANNOYING AND nil DOES NOT ALWAYS EQUAL nil.
+		// IT IS MUCH SAFER TO ONLY SET HasLayerIndex TO layerIdx WHEN WE KNOW FOR
+		// SURE layerIdx != nil!!!
```

Comment:
> GO CAN SOMETIMES BE ANNOYING AND nil DOES NOT ALWAYS EQUAL nil.

The way this comment goes by sounds more like "don't touch this, otherwise things might break!". But if it told us a bit more how things will break if we directly assign the `nil` interface type it would be appreciate for future readers.

Why does this only affect `HasLayerIndex`, for example? Do we have other potential places where we need to avoid setting an interface with a non-nil type and a nil value?

Reply:
Fair. I'll link to an article which explains it more

> Do we have other potential places where we need to avoid setting an interface with a non-nil type and a nil value?

I hope not, but I haven't checked the whole codebase. I do know for this particular spot, this is the only place.

This mostly comes up when dealing with protos and `oneof` fields

Reply:
[886f7e9](https://github.com/stackrox/stackrox/pull/12820/commits/886f7e91ee821bfa0cc08631d6d08bd5cffe3ef3)

Decided to both keep the original all-caps warning and add a better explanation. I figure the all-caps warning makes it clear "DO NOT TOUCH THIS"